modules/src/applications/transfer/denom.rs

```diff
@@ -0,0 +1,653 @@
+            IbcCoin::Hashed(coin) => write!(f, "{}-{}", coin.amount, coin.denom),
+            IbcCoin::Base(coin) => write!(f, "{}-{}", coin.amount, coin.denom),
+        }
+    }
+}
+
+impl TryFrom<RawCoin> for IbcCoin {
+    type Error = Error;
+
+    fn try_from(proto: RawCoin) -> Result<IbcCoin, Self::Error> {
+        if proto.denom.starts_with(IBC_DENOM_PREFIX) {
+            Ok(Self::Hashed(HashedCoin::try_from(proto)?))
+        } else {
+            Ok(Self::Base(BaseCoin::try_from(proto)?))
+        }
+    }
+}
+
+impl From<IbcCoin> for RawCoin {
+    fn from(ibc_coin: IbcCoin) -> Self {
+        let (denom, amount) = match ibc_coin {
+            IbcCoin::Hashed(c) => (c.denom.to_string(), c.amount.to_string()),
+            IbcCoin::Base(c) => (c.denom.to_string(), c.amount.to_string()),
+        };
+        Self { denom, amount }
+    }
+}
+
+impl From<PrefixedCoin> for IbcCoin {
+    fn from(prefixed_coin: PrefixedCoin) -> Self {
+        let Coin { denom, amount } = prefixed_coin;
+        if !denom.trace_path.is_empty() {
+            IbcCoin::Hashed(HashedCoin {
+                denom: denom.hashed(),
+                amount,
+            })
+        } else {
+            IbcCoin::Base(BaseCoin {
+                denom: denom.base_denom,
+                amount,
+            })
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_denom_validation() -> Result<(), Error> {
+        assert!(Denom::from_str("").is_err(), "empty base denom");
+        assert!(Denom::from_str("uatom").is_ok(), "valid base denom");
+        assert!(DenomTrace::from_str("").is_err(), "empty denom trace");
+        assert!(
+            DenomTrace::from_str("transfer/channel-0/").is_err(),
+            "empty base denom with trace"
+        );
+        assert!(DenomTrace::from_str("/uatom").is_err(), "empty prefix");
+        assert!(DenomTrace::from_str("//uatom").is_err(), "empty ids");
+        assert!(DenomTrace::from_str("transfer/").is_err(), "single trace");
+        assert!(
+            DenomTrace::from_str("transfer/atom").is_err(),
+            "single trace with base denom"
+        );
+        assert!(
+            DenomTrace::from_str("transfer/channel-0/uatom").is_ok(),
+            "valid single trace info"
+        );
+        assert!(
+            DenomTrace::from_str("transfer/channel-0/transfer/channel-1/uatom").is_ok(),
+            "valid multiple trace info"
+        );
+        assert!(
+            DenomTrace::from_str("(transfer)/channel-0/uatom").is_err(),
+            "invalid port"
+        );
+        assert!(
+            DenomTrace::from_str("transfer/(channel-0)/uatom").is_err(),
+            "invalid channel"
+        );
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_denom_trace() -> Result<(), Error> {
+        assert_eq!(
+            DenomTrace::from_str("transfer/channel-0/uatom")?,
+            DenomTrace {
+                trace_path: "transfer/channel-0".parse()?,
```

Comment:
Can we write the expected values explicitly (e.g. `vec![TracePath(vec![TracePrefix::new(...)])]`) as opposed to using `parse()`? In the current form, it's hard for me to tell whether or not we're comparing `DenomTrace::from_str` with itself (or almost).

This applies to all other expected values in the tests